File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        left, right = 0, 0
        n = len(s)

        def expand(l, r):
            while l - 1 >= 0 and r + 1 < n and s[l - 1] == s[r + 1]:
                l -= 1
                r += 1
            return l, r

        for i in range(n - 1):
            l, r = expand(i, i)
            if r - l > right - left:
                left, right = l, r
            if s[i] == s[i + 1]:
                l, r = expand(i, i + 1)
                if r - l > right - left:
                    left, right = l, r
        
        return s[left:(right + 1)]

File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.cpp
class Solution {
public:
    string longestPalindrome(string s) {
        int opt_left = 0, opt_right = 0;
        for (int i = 0; i < s.size(); i++) {
            vector<int> pair = expand(s, i, i);
            if (pair[1] - pair[0] > opt_right - opt_left) {
                opt_left = pair[0];
                opt_right = pair[1];
            }
            if (i + 1 < s.size() && s[i + 1] == s[i]) {
                vector<int> pair = expand(s, i, i + 1);
                if (pair[1] - pair[0] > opt_right - opt_left) {
                    opt_left = pair[0];
                    opt_right = pair[1];
                }
            }
        }
        return s.substr(opt_left, opt_right - opt_left + 1);
    }

private: 
    vector<int> expand(string s, int left, int right) {
        while (left - 1 >= 0 && right + 1 < s.size() && s[left - 1] == s[right + 1]) {
            left--;
            right++;
        }
        return {left, right};
    }   
};